aquadoggo/src/network/service.rs

```diff
@@ -79,177 +85,388 @@ pub async fn network_service(
                         }
                     }
-                },
-                SwarmEvent::Behaviour(BehaviourEvent::Ping(Event { peer, result: _ })) => {
-                    debug!("Ping from: {peer}")
                 }
-                SwarmEvent::ConnectionClosed {
-                    peer_id,
-                    endpoint,
-                    num_established,
-                    cause,
-                } => {
-                    info!("ConnectionClosed: {peer_id} {endpoint:?} {num_established} {cause:?}")
+                Some(ConnectionError::KeepAliveTimeout) => {
+                    debug!("Connection timed out with peer {peer_id}");
+                }
+                Some(ConnectionError::Handler(_)) => {
+                    warn!("Connection handler error occurred with peer {peer_id}");
+                }
+                None => {
+                    debug!("Connection closed with peer {peer_id}");
+                }
+            },
+            SwarmEvent::ExpiredListenAddr {
+                listener_id,
+                address,
+            } => trace!("Expired listen address: {listener_id:?} {address}"),
+            SwarmEvent::IncomingConnection {
+                local_addr,
+                send_back_addr,
+            } => trace!("Incoming connection: {local_addr} {send_back_addr}"),
+            SwarmEvent::IncomingConnectionError {
+                local_addr,
+                send_back_addr,
+                error,
+            } => {
+                warn!("Incoming connection error occurred with {local_addr} and {send_back_addr}: {error}");
+            }
+            SwarmEvent::ListenerClosed {
+                listener_id,
+                addresses,
+                reason,
+            } => trace!("Listener closed: {listener_id:?} {addresses:?} {reason:?}"),
+            SwarmEvent::ListenerError { error, .. } => {
+                warn!("Listener failed with error: {error}")
+            }
+            SwarmEvent::NewListenAddr {
+                address,
+                listener_id: _,
+            } => {
+                debug!("Listening on {address}");
+            }
+            SwarmEvent::OutgoingConnectionError { peer_id, error } => match peer_id {
+                Some(id) => {
+                    warn!("Outgoing connection error with peer {id} occurred: {error}");
+                }
+                None => {
+                    warn!("Outgoing connection error occurred: {error}");
+                }
+            },
+
+            // ~~~~
+            // mDNS
+            // ~~~~
+            SwarmEvent::Behaviour(BehaviourEvent::Mdns(event)) => match event {
+                mdns::Event::Discovered(list) => {
+                    for (peer_id, multiaddr) in list {
+                        debug!("mDNS discovered a new peer: {peer_id}");
+
+                        if let Err(err) = self.swarm.dial(multiaddr) {
```

Comment:
I fixed this to dial one multiaddr at a time, this reduces duplicate outbound connections.